python/cudf_polars/cudf_polars/containers/dataframe.py

```diff
@@ -0,0 +1,175 @@
+# SPDX-License-Identifier: Apache-2.0
+
+"""A dataframe, with some properties."""
+
+from __future__ import annotations
+
+from functools import cached_property
+from typing import TYPE_CHECKING
+
+import polars as pl
+
+import cudf._lib.pylibcudf as plc
+
+from cudf_polars.containers.column import Column
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping, Sequence, Set
+
+    from typing_extensions import Self
+
+    import cudf
+
+    from cudf_polars.containers.scalar import Scalar
+
+
+__all__: list[str] = ["DataFrame"]
+
+
+class DataFrame:
+    """A representation of a dataframe."""
+
+    columns: list[Column]
+    scalars: list[Scalar]
+    table: plc.Table | None
+
+    def __init__(self, columns: Sequence[Column], scalars: Sequence[Scalar]) -> None:
+        self.columns = list(columns)
+        self._column_map = {c.name: c for c in self.columns}
+        self.scalars = list(scalars)
+        if len(scalars) == 0:
+            self.table = plc.Table([c.obj for c in columns])
+        else:
+            self.table = None
+
+    def copy(self) -> Self:
+        """Return a shallow copy of self."""
+        return type(self)(self.columns, self.scalars)
+
+    def to_polars(self) -> pl.DataFrame:
+        """Convert to a polars DataFrame."""
+        assert len(self.scalars) == 0
+        return pl.from_arrow(
+            plc.interop.to_arrow(
+                self.table,
+                [plc.interop.ColumnMetadata(name=c.name) for c in self.columns],
+            )
+        )
+
+    @cached_property
+    def column_names_set(self) -> frozenset[str]:
+        """Return the column names as a set."""
+        return frozenset(c.name for c in self.columns)
+
+    @cached_property
+    def column_names(self) -> list[str]:
+        """Return a list of the column names."""
+        return [c.name for c in self.columns]
+
+    @cached_property
+    def num_columns(self) -> int:
+        """Number of columns."""
+        return len(self.columns)
+
+    @cached_property
+    def num_rows(self) -> int:
+        """Number of rows."""
+        if self.table is None:
+            raise ValueError("Number of rows of frame with scalars makes no sense")
+        return self.table.num_rows()
+
+    @classmethod
+    def from_cudf(cls, df: cudf.DataFrame) -> Self:
+        """Create from a cudf dataframe."""
+        return cls(
+            [Column(c.to_pylibcudf(mode="read"), name) for name, c in df._data.items()],
+            [],
+        )
+
+    @classmethod
+    def from_table(cls, table: plc.Table, names: Sequence[str]) -> Self:
+        """Create from a pylibcudf table."""
+        # TODO: strict=True when we drop py39
+        if table.num_columns() != len(names):
+            raise ValueError("Mismatching name and table length.")
+        return cls([Column(c, name) for c, name in zip(table.columns(), names)], [])
+
+    def with_sorted(self, *, like: DataFrame, subset: Set[str] | None = None) -> Self:
```

Comment:
If we rename the column version probably should rename this too for consistency.